src/core/net/ip6_mpl.cpp

```diff
@@ -105,38 +102,265 @@ ThreadError Mpl::ProcessOption(const Message &aMessage, const Address &aAddress,
+
+exit:
+
+    if (shouldIncrement)
+    {
+        mSequence++;
+    }
+
+    return error;
+}
+
+ThreadError Mpl::ProcessOption(Message &aMessage, const Address &aAddress)
+{
+    ThreadError error;
+    OptionMpl option;
+
+    VerifyOrExit(aMessage.Read(aMessage.GetOffset(), sizeof(option), &option) >= OptionMpl::kMinLength &&
+                 (option.GetSeedIdLength() == OptionMpl::kSeedIdLength0 ||
+                  option.GetSeedIdLength() == OptionMpl::kSeedIdLength2),
+                 error = kThreadError_Drop);
+
+    if (option.GetSeedIdLength() == OptionMpl::kSeedIdLength0)
+    {
+        // Retrieve MPL Seed Id from the IPv6 Source Address.
+        option.SetSeedId(HostSwap16(aAddress.mFields.m16[7]));
+    }
+
+    // Check MPL Data Messages in the MPL Buffered Set against sequence number.
+    UpdateBufferedSet(option.GetSeedId(), option.GetSequence());
 
-    entry->mSeed = option.GetSeed();
-    entry->mSequence = option.GetSequence();
-    entry->mLifetime = kLifetime;
-    mTimer.Start(1000);
+    // Check if the MPL Data Message is new.
+    SuccessOrExit(error = UpdateSeedSet(option.GetSeedId(), option.GetSequence()));
+
+    // Enter the MPL Data Message in the MPL Buffered Message Set.
+    AddBufferedMessage(aMessage, option.GetSeedId(), option.GetSequence(), true);
 
 exit:
     return error;
 }
 
-void Mpl::HandleTimer(void *aContext)
+void Mpl::HandleRetransmissionTimer(void *aContext)
+{
+    static_cast<Mpl *>(aContext)->HandleRetransmissionTimer();
+}
+
+void Mpl::HandleRetransmissionTimer()
+{
+    uint32_t now = Timer::GetNow();
+    uint32_t nextDelta = 0xffffffff;
+    MplBufferedMessageMetadata messageMetadata;
+
+    Message *message = mBufferedMessageSet.GetHead();
+    Message *nextMessage = NULL;
+
+    while (message != NULL)
+    {
+        nextMessage = message->GetNext();
+        messageMetadata.ReadFrom(*message);
+
+        if (messageMetadata.IsLater(now))
+        {
+            // Calculate the next retransmission time and choose the lowest.
+            if (messageMetadata.GetTransmissionTime() - now < nextDelta)
+            {
+                nextDelta = messageMetadata.GetTransmissionTime() - now;
+            }
+        }
+        else
+        {
+            // Update number of transmission timer expirations.
+            messageMetadata.SetTransmissionCount(messageMetadata.GetTransmissionCount() + 1);
+
+            if (messageMetadata.GetTransmissionCount() < GetTimerExpirations())
+            {
+                Message *messageCopy = message->Clone(message->GetLength() - sizeof(MplBufferedMessageMetadata));
+
+                if (messageCopy != NULL)
+                {
+                    mIp6.EnqueueDatagram(*messageCopy);
+                }
+
+                // Emulate Trickle timer behavior and set up the next retransmission within [0,I] range.
+                messageMetadata.SetTransmissionTime(now + (otPlatRandomGet() % (kDataMessageInterval + 1)));
```

Comment:
If we want to be closer to Trickle behavior, the random period starts at the beginning of the next interval.

nit: I don't think the `+ 1` is necessary.  Thread defines the interval as `[0, I)`.